transformers.py

```diff
@@ -93,14 +91,24 @@ def __call__(self, inputs):
     inputs: [B, S, D_m]
     returns: [B, S, D_m]
     """
-    # MHA(Q, K, V) = Concat(head_1...head_h)W^O
-    heads = []
-    for i in range(self.num_heads):
-      # head_i = Attention(QW_i^Q, KW_i^K, VW_i^V)
-      head_i = attention(self.W_Q[i](inputs), self.W_K[i](inputs), self.W_V[i](inputs)) # [B, S, D_m/h]
-      heads.append(head_i)
-    multi_head = self.W_O(tf.concat(heads, -1))  # [B, S, D_m]
-    return multi_head
+    B, S, D_m = tf.unstack(tf.shape(inputs))
+    H = self.num_heads
+    P = self.projection_size
+    O = self.output_size
+
+    qkv = self.W_qkv(inputs)  # [B, S, 3 * O]
+    qkv = tf.reshape(qkv, [B, S, H, 3, P])  # unmerge heads, QKV, and projections
+    qkv = tf.transpose(qkv, [0, 2, 1, 3, 4])  # [B, H, S, 3, P]
+    qkv = tf.reshape(qkv, [B * H, S, 3, P])  # merge heads into batch dim
```

Comment:
Clever that you can do this...I guess head is just a batch-like dimension since it's indicating how large of a chunk to process for the input